### PR TITLE
Redaction

### PR DESCRIPTION
**Pull request by BrandonSharratt**
*Issue:*https://github.com/bcgov/entity/issues/21607
*Description of changes:*
UI passes user account id for checking product access. Backend uses product access to determine if user is CA and role to determine if user is staff, BTR Access document is followed for how information should be redacted

Caching is not handled by this PR see #22364

Unclear how to handle the fact that the user won't be able to easily edit and resubmit if they are not a CA (won't see all the information)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the namex license (Apache 2.0).

### CODE REVIEW

File: btr-web/btr-main-app/services/file-significant-individual.ts [hunk 1]
```diff
@@ -152,8 +152,8 @@ const submitSignificantIndividualFiling = async (sif: SignificantIndividualFilin
   return { data: data.value, error: error.value }
 }
 
-const getBtrFiling = async (businessIdentifier: string) => {
-  const url = `${constructBtrApiURL()}/plots/entity/${businessIdentifier}`
+const getBtrFiling = async (businessIdentifier: string, account?: string) => {
+  const url = `${constructBtrApiURL()}/plots/entity/${businessIdentifier}${account ? `?account_id=${account}` : ''}`
```
Comment (kialj876) [inline btr-web/btr-main-app/services/file-significant-individual.ts:156]: please remove these UI changes - the account id is already added into the header of the request

File: btr-api/src/btr_api/resources/submission.py [hunk 2]
```diff
@@ -40,18 +40,132 @@
...
 bp = Blueprint("submission", __name__)
 
+REDACT_MONONYM = "mono"
+REDACT_MONONYM_FN = "mono_fn"
+REDACT_EMAIL = "mono_email"
+REDACT_PHONE = "mono_phone"
+REDACT_FULL = "full"
+REDACT_EMPTY = "empty"
+REDACT_DATE = "date"
+REDACT_IDENTIFIER = "identifier"
```
Comment (kialj876) [inline btr-api/src/btr_api/resources/submission.py:74]: I think it probably makes sense to put these in a REDACT enum?

File: btr-api/src/btr_api/resources/submission.py [hunk 3]
```diff
@@ -40,18 +40,132 @@
...
+        return "***"
+    # This is a space instead of blank because if it's empty the ui shows undefined in some spots
+    elif redactType == REDACT_EMPTY:
+        return " "
+    elif redactType == REDACT_DATE:
+        return field.split("-")[0]
+    elif redactType == REDACT_IDENTIFIER:
+        return "*** **" + field[6:]
+
+    return field
```
Comment (kialj876) [inline btr-api/src/btr_api/resources/submission.py:167]: pls move the redact code into a util or the submission service

File: btr-api/src/btr_api/resources/submission.py [hunk 4]
```diff
@@ -75,10 +189,12 @@ def get_entity_submission(business_identifier: str):
 
     try:
         btr_auth.is_authorized(request=request, business_identifier=business_identifier)
+        account_id = request.args.get("account_id")
```
Comment (kialj876) [inline btr-api/src/btr_api/resources/submission.py:192]: grab this from the header please. You can see an example inside 'is_authorized'

File: btr-api/src/btr_api/services/auth.py [hunk 5]
```diff
@@ -59,66 +65,136 @@ def __init__(self, app: Flask = None):
...
+                error = f"{resp.status_code} - {str(resp.json())}"
+                self.app.logger.debug("Invalid response from auth-api: %s", error)
+                request.product_resp = {}
+
+            if resp.status_code != HTTPStatus.OK:
+                error = f"Unauthorized access to products {account_id}"
+                self.app.logger.debug(error)
+                request.product_resp = {}
+
+            request.product_resp = resp
```
Comment (kialj876) [inline btr-api/src/btr_api/services/auth.py:169]: I see your adding stuff to the request object and then passing that around, but it's hard to follow.

I don't think it adds a lot of value to add stuff to the request obj and pass it around everywhere (especially when its renamed to other things), but we can discuss

**Review by kialj876 — COMMENTED** [review]
Hey brandon, I've got some comments. Let me know if you would like to meet

File: btr-api/src/btr_api/services/auth.py [hunk 6]
```diff
@@ -59,66 +65,136 @@ def __init__(self, app: Flask = None):
...
+            self.app.logger.info(products)
+            roles = user.auth_resp.json()["roles"]
+            if any(
+                product
+                for product in products
+                if product["code"].upper() == self.CA_PRODUCT and product["subscriptionStatus"].upper() == "ACTIVE"
+            ):
+                self.app.logger.info("Get User Type: " + self.USER_COMPETENT_AUTHORITY)
+                return self.USER_COMPETENT_AUTHORITY
+            if any(role for role in roles if role == self.STAFF_ROLE):
```
[the remark this answers is not in the document]
Comment (kialj876) [reply btr-api/src/btr_api/services/auth.py:97]: Okay I was wrong about this, we will use the JWT info for verification, but we should use the common jwt.validate_roles method here

File: btr-api/src/btr_api/services/auth.py [hunk 7]
```diff
@@ -66,59 +72,129 @@ def init_app(self, app: Flask):
...
+                HTTPStatus.SERVICE_UNAVAILABLE,
+                [{'message': 'Unable to get service account token.', 'reason': err.with_traceback(None)}],
+            ) from err
 
     def get_authorization_header(self, request) -> str:
         """Gets authorization header from request."""
         authorization_header = request.headers.get('Authorization', None)
         if not authorization_header:
-            error = f'Missing authorization header: {request.headers}'
+            error = f"Missing authorization header: {request.headers}"
```
Comment (hfekete) [inline btr-api/src/btr_api/services/auth.py:137]: styling: keeping the quotes same through out the code

File: btr-api/src/btr_api/resources/submission.py (hunk 3 above)
Comment (hfekete) [reply btr-api/src/btr_api/resources/submission.py:167]: +1 Yes, please put it in the util, I think the best is if service gets all data as is, and the data can be obfuscated wherever it needs to be.

File: btr-api/src/btr_api/services/auth.py [hunk 8]
```diff
@@ -66,59 +72,129 @@ def init_app(self, app: Flask):
...
         return authorization_header
 
+    # TODO: Add caching
+    def product_authorizations(self, request, account_id: str) -> bool:
+        """Authorize the user for access to the service."""
+        if not account_id:
+            error = 'Missing account_id'
+            self.app.logger.debug(error)
+            request.product_resp = {}
+            return True
```
Comment (hfekete) [inline btr-api/src/btr_api/services/auth.py:150]: is it authorized if there is no account_id ?
What does return value do ?

File: btr-api/src/btr_api/services/auth.py (hunk 5 above)
Comment (hfekete) [reply btr-api/src/btr_api/services/auth.py:169]: also, it looks like it will override it even on the error, if any of the error response ifs is triggered, it will set this value to empty, then after the ifs, it will set it to resp itself, is this intended

File: btr-web/btr-main-app/services/file-significant-individual.ts (hunk 1 above)
Comment (hfekete) [reply btr-web/btr-main-app/services/file-significant-individual.ts:156]: Sorry, that was my mistake, I suggested him this. Is there a reason why we should use the one from the headers, rather then using the path and/or query params?

To me it looks like it could be kinda messy to have some data in headers, some data in path.

Though, I could understand that authorization data stay in header.

File: btr-api/src/btr_api/resources/submission.py [hunk 9]
```diff
@@ -40,20 +40,134 @@
...
+                rv += ' '
+            rv += word[0:1] + '***'
+        return rv
+    elif redactType == REDACT_EMAIL:
+        return field.split('@')[0][0:1] + '***' + '@***.' + field.split('.')[-1]
+    elif redactType == REDACT_PHONE:
+        return field[0:-7] + '***'
+    elif redactType == REDACT_FULL:
+        return '***'
+    # This is a space instead of blank because if it's empty the ui shows undefined in some spots
```
Comment (hfekete) [inline btr-api/src/btr_api/resources/submission.py:159]: keep it as is, but maybe we need to discuss this. As it might get messy if we keep putting spaces in the values.

File: btr-web/btr-main-app/services/file-significant-individual.ts (hunk 1 above)
Comment (kialj876) [reply btr-web/btr-main-app/services/file-significant-individual.ts:156]: we use the account id in all the apis from the header and we do it already in the btr-api. I see no reason to change that flow. It's also not part of the path, it's being added as an arg here

File: btr-api/src/btr_api/services/auth.py (hunk 8 above)
Comment (BrandonSharratt) [reply btr-api/src/btr_api/services/auth.py:145]: The comment is perhaps misleading as the function doesn't restrict access it just sets the products I'll update the comment

File: btr-web/btr-main-app/pages/ownerChange.vue [hunk 10]
```diff
@@ -125,7 +126,7 @@ onBeforeMount(async () => {
   // FUTURE: put in a loading page or something while this happens in case network is slow
   await useBcrosBusiness().loadBusiness(identifier)
   const siControlStore = useSiControlStore()
-  const { data, error } = await fileSIApi.getBtrFiling(identifier)
+  const { data, error } = await fileSIApi.getBtrFiling(identifier, account.currentAccount?.id)
```
Comment (kialj876) [inline btr-web/btr-main-app/pages/ownerChange.vue:129]: forgot to remove changes in here

**Review by kialj876 — APPROVED** [review]
Nice work, thank you for the updates. One last thing comment I noted before merging

File: btr-api/src/btr_api/services/auth.py (hunk 7 above)
Comment (BrandonSharratt) [reply btr-api/src/btr_api/services/auth.py]: This is resolved
